### DISCUSSION TITLE
when to use toString() and valueOf()?

### DISCUSSION THREAD

**Discussion by wjm1314** [denysdovhan/wtfjs#189] [WTF Questions] [answered]
### _Example(Unexpected result)_:
**var foo = {
    toString: function() { return 'foo'},
    valueOf: function() { return 5}
}
console.log(foo + 'bar');  //5bar
console.log([foo,'bar'].join(''));  //foobar**

**Comment by pero5ar** [comment, marked as answer]
The `join` is actually perfectly expected. You're joining array elements into a string so it triggers `toString` for each element. 

The `+` is also expected, although a bit less, the thing is that `+` is primarily an arithmetic operator so it triggers `valueOf`, however the `valueOf` a string primitive is that string primitive and since one of the values in the expression remained a string the `+` falls back to being a concatenation operator. This can be seen as a bit WTF, however situations similar to that are covered in the **Funny math** section.

**Comment by wjm1314** [comment]
oh~~~, I got it! Thank you!
